### DISCUSSION TITLE
Error when enrolling public key: "Cant share car"

### DISCUSSION THREAD

**Discussion by JensBiltoft** [llamafilm/tesla-http-proxy-addon#101] [Q&A] [answered]
Hi

I have now tried for several days to get this to work, and I have managed to get to the point where I try to "Enroll public key in your vehicle" When I press this and the Tesla App launches I get an error message 

"Unable to share Vehicle
This third-party isn't registered with Tesla. We can't grant them access at this time" (roughly translated from Danish)

I use duckdns and NGINX and I have followed the guide as far as I can tell, but I must have done something wrong.

The log from duckdns
```
Renew certificate for domains: biltofts.duckdns.org
tesla-biltofts.duckdns.org and aliases: 
# INFO: Using main config file /data/workdir/config
Processing biltofts.duckdns.org with alternative names: biltoftstesla.duckdns.org
 + Checking domain name(s) of existing cert... unchanged.
 + Checking expire date of existing cert...
 + Valid till Aug 24 11:38:23 2024 GMT (Longer than 30 days). Skipping renew!
```

The log from NGINX
```
INFO: Running nginx...
2024/05/27 18:27:27 [error] 150#150: *1 upstream prematurely closed connection while reading response header from upstream, client: 192.168.0.1, server: biltoftstesla.duckdns.org, request: "GET /.well-known/appspecific/com.tesla.3p.public-key.pem HTTP/2.0", upstream: "http://172.30.32.1:8123/.well-known/appspecific/com.tesla.3p.public-key.pem", host: "biltoftstesla.duckdns.org"
```

Log from tesla-http-proxy-addon
```
 * Running on all addresses (0.0.0.0)
 * Running on http://127.0.0.1:8099
 * Running on http://172.30.33.4:8099
[18:27:25] werkzeug:INFO: Press CTRL+C to quit
[18:27:26] INFO: Found existing keypair
[18:27:26] INFO: Testing tesla-biltofts.duckdns.org for an associated IP address...
tesla-biltofts.duckdns.org has address 85.184.140.111
tesla-biltofts.duckdns.org mail is handled by 50 biltoftstesla.duckdns.org.
[18:27:27] INFO: Found an IP address for biltoftstesla.duckdns.org
[18:27:27] INFO: Testing public key...
HTTP/2 502 
server: nginx
date: Mon, 27 May 2024 16:27:27 GMT
content-type: text/html
content-length: 150
strict-transport-security: max-age=31536000; includeSubDomains

[18:27:27] ERROR: HTTP status code 502; Sleeping 2 minutes before retry.
[18:29:27] ERROR: HTTP status code 502; Sleeping 2 minutes before retry.
```

If I try to open the page 
```
https://biltoftstesla.duckdns.org/.well-known/appspecific/com.tesla.3p.public-key.pem
```
in Chrome, I get a big fat warning that it is not safe, and if I proceed I get the "502 Bad Gateway"

![image](https://github.com/llamafilm/tesla-http-proxy-addon/assets/144684192/9a3335b6-2767-47fa-b4d2-26d4ef046365)

Please, can someone tell me what I am doing wrong?
/Jens

**Comment by JensBiltoft** [comment]
I see there is a simillar problem here : [Error when enrolling vehicle #32](https://github.com/llamafilm/tesla-http-proxy-addon/discussions/32)

Unfortunately this does not solve the problem for me. 

My problem is something else.

I now see these logs in the addon log
```
* Running on all addresses (0.0.0.0)
 * Running on http://127.0.0.1:8099
 * Running on http://172.30.33.4:8099
[11:15:16] werkzeug:INFO: Press CTRL+C to quit
[11:15:17] INFO: Found existing keypair
[11:15:17] INFO: Testing public key...
curl: (35) OpenSSL/3.1.5: error:0A000458:SSL routines::tlsv1 unrecognized name
Warning: Problem (retrying all errors). Will retry in 1 seconds. 14 retries 
Warning: left.
```
These continue for 15 minutes or so, before the addon turns itself off

In the NGINX I sometimes see logs like this
```
2024/05/30 09:53:38 [error] 150#150: *78 connect() failed (111: Connection refused) while connecting to upstream, client: 185.181.22.4, server: biltofts.duckdns.org, request: "GET /api/websocket HTTP/1.1", upstream: "http://172.30.32.1:8123/api/websocket", host: "biltofts.duckdns.org"
```

**Reply by JensBiltoft** [reply, marked as answer]
I finally managed to get the addon to work, and get my car to accept the key.

I had made several mistakes in my first attempts so I made a fresh start, where I deleted the duckdns, NGINX and tesla http proxy addons, and deleted /ssl folder and the /share folder

Then I created a new account at developer.tesla.com and created a new domain at duckdns.

Made sure that the new domain had propagated using dns checker.

Then I installed duckdns and NGINX, and verified that duckdns could pass the dns challange (this failed previously, but after reinstall, it passed)

Then I installed Tesla http proxy addon and configured it excactly as described in docs and wiki page [Setup-guide-using-existing-DuckDNS-&-NGINX-Home-Assistant-SSL-proxy](https://github.com/llamafilm/tesla-http-proxy-addon/wiki/Setup-guide-using-existing-DuckDNS-&-NGINX-Home-Assistant-SSL-proxy)

When I create the new account I used Tesla Tax number (not sure it matters)

On the Tesla HTTP proxy Web Interface, I found out that I had to login with my original account, and to grant access for my new account.

Then I got a token, and when I enrolled this in the vehicle it succeeded.

**Comment by mstephan76** [comment]
I have the same problem, I can not enroll the public key in my vehicle... very bad documentation... maybe anyone can make a detailed video with a working workflow or describe every step...
I used HomeAssistant with the old access and I worked perfect... 

1) I have generated a ssl certificat for my sub-domain... 
2) I registered at tesla on den Europe site with VAT Code
3) I could login in HomeAssistant on the Add-ons page to my tesla account
4) I copied the callback URL to the field and generated a token
At this point nothing more worked... I couln't enroll a public key to my vehicle...

I think on the callback site should/must exist a certificate or file, mustn't it?
Can anyone give an advice or feedback?
Thanks

**Reply by iainbullock** [reply]
'very bad documentation'

Really? Do you think that comment will help you when people do this for free in their spare time?

**Reply by mstephan76** [reply]
Maybe it is a feedback to give the professionals a hint, that there are people who are not so fit in configuring this and they make a better documentation

**Comment by aalllop** [comment]
I have got a similar issue. I'm able to get https://xxxxxxxxx.duckdns.org/.well-known/appspecific/com.tesla.3p.public-key.pem.
 I have configured Tesla HTTP Proxy with may access code and secret using my duckdns domain. When I start Tesla HTTP Proxy I get this warnings on log: 
[13:17:22] webui:INFO: Starting Flask server for Web UI...
[13:17:22] werkzeug:INFO: WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
 * Running on all addresses (0.0.0.0)
 * Running on http://127.0.0.1:8099
 * Running on http://172.30.33.12:8099
[13:17:22] werkzeug:INFO: Press CTRL+C to quit
[13:17:24] INFO: Found existing keypair
[13:17:24] INFO: Testing public key...
curl: (28) Resolving timed out after 2001 milliseconds
Warning: Problem : timeout. Will retry in 1 seconds. 14 retries left.
curl: (60) SSL certificate problem: self-signed certificate
More details here: https://curl.se/docs/sslcerts.html

curl failed to verify the legitimacy of the server and therefore could not
establish a secure connection to it. To learn more about this situation and
how to fix it, please visit the web page mentioned above.
Warning: Problem (retrying all errors). Will retry in 2 seconds. 13 retries 
Warning: left.
curl: (60) SSL certificate problem: self-signed certificate
More details here: https://curl.se/docs/sslcerts.html

My big problem is that I always get the "Unable to share Vehicle
This third-party isn't registered with Tesla. We can't grant them access at this time" when I select "Enroll public key in your vehicle".

At the same time on the log I can see that:

[13:31:30] webui:WARNING: Obtained refresh token: EU_e4b78f264e1dfaaXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXx
[13:31:30] werkzeug:INFO: 172.30.32.2 - - [09/Jun/2024 13:31:30] "GET /callback?callback_url=https://XXXXXXXX.duckdns.org/callback?code%3DEU_0420584cc6ced0620d42ee24fd05827e5e6603455ef9e238989bb1a650f6%26state%3D6bcdedec4d%26issuer%3Dhttps%253A%252F%252Fauth.tesla.com%252Foauth2%252Fv3 HTTP/1.1" 200 -

What is wrong? 

I have also read this post https://github.com/teslamotors/vehicle-command/issues/203 but I am owner of my car, not leasing.

**Comment by JensBiltoft** [comment]
Looks like the same issue I had. Not sure what my problem was, but I saw that duckdns failed on dns challenge on the new domain. So I created a new, and then it passed.

⁣Hent BlueMail til Android ​

Den 9. jun. 2024 13.35, fra 13.35, aalllop ***@***.***> skrev:

**Comment by aalllop** [comment]
Have you créated  another domain only for this plugging? With his proper certificate? I have another question. I have nginx ssl proxy running on 443 port so I could not put tesla proxy on the same port . What port have you used for tesla proxy?

**Reply by baylanger** [reply]
> Have you créated another domain only for this plugging? With his proper certificate?

If you followed the instructions, this is a requirement.

> I have another question. I have nginx ssl proxy running on 443 port so I could not put tesla proxy on the same port . What port have you used for tesla proxy?

Nginx is configured to forward Tesla traffic to the proxy that runs on another port. This configuration is automatically added when you install the proxy.

**Comment by JensBiltoft** [comment]
I only use 443. 
I access my HA installation on myHaDomain.duckdns.org and use anotherDomain.duckdns.org for the http proxy.


⁣Hent BlueMail til Android ​

Den 9. jun. 2024 14.24, fra 14.24, aalllop ***@***.***> skrev:

**Comment by aalllop** [comment]
I have created another dns url but no luck

**Comment by aalllop** [comment]
I have found the problem but not the solution.  My ha is resolving de my_teala.fuckdns.org to the Internet ip so is not able to get de certificate. I have configured ha dns to the local dns installed on dnsmasq but is resolving dns from internet. Any idea ?

**Comment by aalllop** [comment]
I have configured on HA the dns to a router on my LAN. On the router I have defined my_tesla duckdns.org to the ip of HA server in my lan and now the tesla http proxy is able to get and register the token on tesla, and I can configure tesla integration
